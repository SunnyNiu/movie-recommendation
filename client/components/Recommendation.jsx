import React from 'react'
import { connect } from 'react-redux'
import { Title, Button, MovieContainer, Img } from '../Recommendation.styles'
import { withRouter } from 'react-router-dom'
import { clearAll, fetchRecommendMoviesNotInShowedBefore } from '../../redux/movie.actions'

class Recommendation extends React.Component {
  componentDidMount () {
    const { fetchRecommendMoviesNotInShowedBefore, moviesId, genres } = this.props
    const arr = Object.values(genres)
    const max = Math.max(...arr)
    const likedGenre = Object.keys(genres).find(key => genres[key] === max)
    fetchRecommendMoviesNotInShowedBefore(likedGenre, moviesId)
  }

  render () {
    const { history, movie: { name, image }, clearAll } = this.props

    return (<MovieContainer>
      <Title>You Probably Like this Movie: {name} </Title>
      <Img src={image} alt='movieImage'/>
      <Button onClick={() => { history.push('/'); clearAll() }}> Back to Home! </Button>
    </MovieContainer>)
  }
}

function mapStateToProps (state) {
  return {
    genres: state.genres,
    movie: state.movie,
    moviesId: state.moviesId
  }
}

const mapDispatchToProps = dispatch => ({
  clearAll: () => dispatch(clearAll()),
  fetchRecommendMoviesNotInShowedBefore: (genre, moviesId) => dispatch(fetchRecommendMoviesNotInShowedBefore(genre, moviesId))
})
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Recommendation))
