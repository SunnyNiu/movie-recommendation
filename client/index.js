import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import reducer from '../redux/movie.reducer'
import { fetchMovie } from '../redux/movie.actions'
import { BrowserRouter } from 'react-router-dom'

import App from './components/App'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const store = createStore(
  reducer,
  composeEnhancers(applyMiddleware(thunk))
)

function render () {
  const root = document.getElementById('app')
  ReactDOM.render(<Provider store={store}><BrowserRouter><App /></BrowserRouter></Provider>, root)
}

render()
store.subscribe(render)
