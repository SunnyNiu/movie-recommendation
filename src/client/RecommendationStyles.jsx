import styled from 'styled-components'

export const Button = styled.button`
  background: transparent;
  border-radius: 3px;
  border: 2px solid palevioletred;
  color: palevioletred;
  margin: 0em 1em;
  font-size: 1rem;
  text-align: center;
  &:hover {
    cursor: pointer;
    color: black;
  }
`
export const MovieContainer = styled.div`
display: flex;
flex-direction: column;
align-items: center;
`

export const Title = styled.h1`
font-style: italic;
font-size: 1rem;
font-family: Andale Mono, monospace;
text-align: center;
align-items: center;
margin: 1em 0em
`
export const MovieList = styled.div`
font-size: 1rem;
margin:1rem;
padding:2px;
align-items: center;
text-align: center;
font-family: Andale Mono, monospace;
`
