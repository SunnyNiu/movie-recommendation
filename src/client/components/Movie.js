import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Grid from './Grid';
import Cell from './Cell';

export const MovieContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Movie = ({ movie }) => {
  const { Name, wTeaser, yUrl } = movie;

  return (
    <MovieContainer>
      <label>Movie:{Name}</label>{' '}
      <Grid columns="420px 1fr">
        <Cell minWidth="450px">
          <iframe
            title={Name}
            src={yUrl}
            width="400"
            height="300"
            frameBorder="0"
            allowFullScreen
          />
        </Cell>
        <Cell minWidth="450px">
          <p>{wTeaser}</p>
        </Cell>
      </Grid>
    </MovieContainer>
  );
};

Movie.propTypes = {
  Name: PropTypes.string,
  wTeaser: PropTypes.string,
  yUrl: PropTypes.string,
  movie: PropTypes.arrayOf(PropTypes.string).isRequired,
};

Movie.defaultProps = {
  Name: '',
  wTeaser: '',
  yUrl: '',
};

export default Movie;
