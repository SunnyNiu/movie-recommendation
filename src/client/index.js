import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore, compose } from 'redux';
import thunk from 'redux-thunk';
import { HashRouter } from 'react-router-dom';
import reducer from './redux/movie.reducer';

// eslint-disable-next-line import/extensions
import App from './components/App';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(reducer, composeEnhancers(applyMiddleware(thunk)));

function render() {
  const root = document.getElementById('app');
  ReactDOM.render(
    <Provider store={store}>
      <HashRouter>
        <App />
      </HashRouter>
    </Provider>,
    root
  );
}

render();
store.subscribe(render);
