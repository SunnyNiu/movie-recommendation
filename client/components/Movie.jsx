import React from 'react'
import { MovieList } from '../Recommendation.styles'
import Grid from './Grid'
import Cell from './Cell'

class Movie extends React.Component {
  render () {
    const { Name, wTeaser, yUrl } = this.props.movie
    return (
      <MovieList>
        <label>Movie:{Name}</label>
        <Grid columns="420px 1fr">
          <Cell minWidth="450px" center={'right'}>
            <iframe src={yUrl}
              width="400" height="300" frameBorder="0" allowFullScreen></iframe>
          </Cell>
          <Cell minWidth="450px">
            <p>{wTeaser}</p>
          </Cell>
        </Grid>
      </MovieList>
    )
  }
}

export default Movie
