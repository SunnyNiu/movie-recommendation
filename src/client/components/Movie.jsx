import React from 'react'
import { MovieList } from '../RecommendationStyles'
import Grid from './Grid'
import Cell from './Cell'

class Movie extends React.Component {
  render () {
    const { Name, wTeaser, yUrl } = this.props.movie
    return (
      <MovieList>
        <label>Movie:{Name}</label>
        <Grid columns="420px 1fr">
          <Cell minWidth="450px">
            <iframe src={yUrl}
              width="400" height="300" frameBorder="0" allowFullScreen></iframe>
          </Cell>
          <Cell minWidth="450px">
            <p>{wTeaser}</p>
          </Cell>
        </Grid>
      </MovieList>
    )
  }
}

export default Movie
