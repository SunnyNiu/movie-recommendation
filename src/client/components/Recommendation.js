import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Button } from './Button.styles';
import { clearAll, fetchRecommendMoviesCreator } from '../redux/movie.actions';
import Movie from './Movie';

export const Title = styled.h2`
  font-style: italic;
  font-size: 1rem;
  align-items: center;
  margin: 1rem, 0rem;
`;

export const MovieContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

class Recommendation extends React.Component {
  // eslint-disable-next-line react/no-deprecated
  componentWillMount() {
    const { likedMovies, fetchRecommendMovies } = this.props;
    fetchRecommendMovies(likedMovies);
  }

  render() {
    // eslint-disable-next-line no-shadow
    const { history, movies, clearAll } = this.props;
    const recommendMovies = movies[0];
    return (
      <MovieContainer>
        <Title>You Probably Like these Movies:</Title>

        {recommendMovies === undefined ? (
          ''
        ) : (
          <div>
            {recommendMovies.map((movie, index) => (
              <Movie key={index} movie={movie} />
            ))}
          </div>
        )}

        <Button
          onClick={() => {
            history.push('/');
            clearAll();
          }}
        >
          {' '}
          Back to Home!{' '}
        </Button>
      </MovieContainer>
    );
  }
}

function mapStateToProps(state) {
  return {
    likedMovies: state.likedMovies,
    movies: state.movies,
  };
}

const mapDispatchToProps = (dispatch) => ({
  clearAll: () => dispatch(clearAll()),
  fetchRecommendMovies: (movies) =>
    dispatch(fetchRecommendMoviesCreator(movies)),
});

Recommendation.propTypes = {
  clearAll: PropTypes.func.isRequired,
  movies: PropTypes.arrayOf(PropTypes.string).isRequired,
  likedMovies: PropTypes.arrayOf(PropTypes.string).isRequired,
  history: PropTypes.func.isRequired,
  fetchRecommendMovies: PropTypes.func.isRequired,
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Recommendation)
);
