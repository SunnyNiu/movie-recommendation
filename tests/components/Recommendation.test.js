import React from 'react'
import { mount } from 'enzyme'
import Recommendation from '../../client/components/Recommendation'
import Movie from '../../client/components/Movie'
import { Button } from '../../client/RecommendationStyles'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router'

jest.mock('react-dom')
const mockStore = configureStore([])

describe('<Recommendation /> component tests', () => {
  let store
  let component
  beforeEach(() => {
    store = mockStore({
      movie: '',
      moviesId: [],
      likedMovies: [],
      movies: []
    })

    store.dispatch = jest.fn()

    component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Recommendation />
        </MemoryRouter>
      </Provider>
    )
  })

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

 
  it('should contains <Button />', () => {
    const expected = true
    const wrapper = mount(<Provider store={store}>
      <MemoryRouter>
        <Recommendation />
      </MemoryRouter>
    </Provider>)

    const actual = wrapper.containsMatchingElement(Button)
    expect(actual).toBe(expected)
  })

  it('should contains <Movie />', () => {
    const expected = true
    const wrapper = mount(<Provider store={store}>
      <MemoryRouter>
        <Recommendation />
      </MemoryRouter>
    </Provider>)

    const actual = wrapper.containsMatchingElement(Movie)
    expect(actual).toBe(expected)
  })
})
