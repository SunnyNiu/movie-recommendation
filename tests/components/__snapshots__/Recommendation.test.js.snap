// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recommendation /> component tests should render with given state from Redux store 1`] = `
<div
  className="MovieContainer-jcnqfJ blwKBe"
>
  <h2
    className="Title-kAzUgf dqyLeU"
  >
    You Probably Like these Movies:
  </h2>
  
  <button
    className="Button-hYXUXp heByZf"
    onClick={[Function]}
  >
     
    Back to Home!
     
  </button>
</div>
`;
